apax/data/input_pipeline.py

```diff
@@ -358,88 +360,82 @@ def next_power_of_two(x):
-            "numbers": [],
-            "n_atoms": [],
-            "positions": [],
-            "box": [],
-            "idx": [],
-            "offsets": [],
+            "numbers": np.zeros((n_samples, max_atoms), dtype=np.int16),
+            "n_atoms": np.zeros(n_samples, dtype=np.int16),
+            "positions": np.zeros((n_samples, max_atoms, 3), dtype=np.float64),
+            "box": np.zeros((n_samples, 3, 3), dtype=np.float32),
         }
 
         labels = {
-            "energy": [],
+            "energy": np.zeros(n_samples, dtype=np.float64),
         }
 
         if self.forces:
-            labels["forces"] = []
-        if self.stress:
-            labels["stress"] = []
-
-        for sample in samples:
-            inp, lab = sample
-
-            inputs["numbers"].append(inp["numbers"])
-            inputs["n_atoms"].append(inp["n_atoms"])
-            inputs["positions"].append(inp["positions"])
-            inputs["box"].append(inp["box"])
-            idx, offsets = compute_nl(inp["positions"], inp["box"], self.cutoff)
-            inputs["idx"].append(idx)
-            inputs["offsets"].append(offsets)
-
-            labels["energy"].append(lab["energy"])
-            if self.forces:
-                labels["forces"].append(lab["forces"])
-            if self.stress:
-                labels["stress"].append(lab["stress"])
+            labels["forces"] = np.zeros((n_samples, max_atoms, 3), dtype=np.float64)
 
-        max_atoms = np.max(inputs["n_atoms"])
-        max_nbrs = np.max([idx.shape[1] for idx in inputs["idx"]])
+        idxs = []
+        offsets = []
+        for i, (inp, lab) in enumerate(samples):
+            inputs["numbers"][i, : inp["n_atoms"]] = inp["numbers"]
+            inputs["n_atoms"][i] = inp["n_atoms"]
+            inputs["positions"][i, : inp["n_atoms"]] = inp["positions"]
+            inputs["box"][i] = inp["box"]
 
-        max_atoms = next_power_of_two(max_atoms)
-        max_nbrs = next_power_of_two(max_nbrs)
+            idx, offset = compute_nl(inp["positions"], inp["box"], self.cutoff)
+            idxs.append(idx)
+            offsets.append(offset)
 
-        for i in range(len(inputs["n_atoms"])):
-            inputs["idx"][i], inputs["offsets"][i] = pad_nl(
-                inputs["idx"][i], inputs["offsets"][i], max_nbrs
-            )
+            labels["energy"][i] = lab["energy"]
+            if self.forces:
+                labels["forces"][i, : inp["n_atoms"]] = lab["forces"]
 
-            zeros_to_add = max_atoms - inputs["numbers"][i].shape[0]
-            inputs["positions"][i] = np.pad(
-                inputs["positions"][i], ((0, zeros_to_add), (0, 0)), "constant"
-            )
-            inputs["numbers"][i] = np.pad(
-                inputs["numbers"][i], (0, zeros_to_add), "constant"
-            ).astype(np.int16)
+        max_nbrs = np.max([idx.shape[1] for idx in idxs])
+        max_nbrs = next_power_of_two(max_nbrs)  # TODO better intervals
```

Comment:
TODO: better intervals?

Reply:
line 424 needs to be deleted. `max_nbrs` gets overwritten in the next line anyway, which is the IMO more sensible approach.
Powers of two are pretty wasteful. The input now allows the next integer (default 2000) to pad neighbors to.

Reply:
oh, that means I need to rerun the benchmark.... Currently, the padding was applied twice. The NL was first padded to the next power of 2 and then to the next integer.

Reply:
so maybe, the pbp dataset will close the performance gap with the cached one afterall.